src/mds/CInode.cc

```diff
@@ -4374,7 +4374,19 @@ void CInode::scrub_initialize(CDentry *scrub_parent,
 			      MDSInternalContextBase *f)
 {
   dout(20) << __func__ << " with scrub_version " << get_version() << dendl;
-  assert(!scrub_is_in_progress());
+  if (scrub_is_in_progress()) {
+    dout(20) << __func__ << " inode moved during scrub, reinitializing "
+	     << dendl;
+    assert(scrub_infop->scrub_parent);
+    CDentry *dn = scrub_infop->scrub_parent;
+    CDir *dir = dn->dir;
+    dn->put(CDentry::PIN_SCRUBPARENT);
+    assert(dir->scrub_infop && dir->scrub_infop->directory_scrubbing);
+    dir->scrub_infop->directories_scrubbing.erase(dn->key());
+    dir->scrub_infop->others_scrubbing.erase(dn->key());
+    delete scrub_infop;
+    scrub_infop = new scrub_info_t();
+  }
```

Comment:
what happens if mds has already started invalidating the inode; or the inode is directory, mds has started scrubbing its dirfrags. I think it's better to keep the scrub data, only update scrub_infop->scrub_parent.

Reply:
We don't want to call scrub_dentry_finished because that records the inode as scrubbed. We definitely don't want to delete scrub_infop, though.

Reply:
I don't see CDir::scrub_dentry_finished() marks inode as scrubbed

Reply:
https://github.com/fullerdj/ceph/blob/wip-djf-22288/src/mds/CDir.cc#L3221 / https://github.com/fullerdj/ceph/blob/wip-djf-22288/src/mds/CDir.cc#L3225